nsxt/resource_nsxt_manager_cluster.go

```diff
@@ -0,0 +1,414 @@
+		ClusterId:                   &clusterID,
+		IpAddress:                   &masterNodeIP,
+		Username:                    &username,
+		Password:                    &password,
+	}
+	_, err = client.Joincluster(joinClusterParams)
+	if err != nil {
+		return fmt.Errorf("Failed to join node to cluster: %s, cluster id: %s", err, clusterID)
+	}
+	return nil
+}
+
+func getHostCredential(m interface{}) (string, string) {
+	nsxtClient := m.(nsxtClients)
+	username := nsxtClient.NsxtClientConfig.UserName
+	password := nsxtClient.NsxtClientConfig.Password
+	return username, password
+}
+
+func resourceNsxtManagerClusterRead(d *schema.ResourceData, m interface{}) error {
+	connector := getPolicyConnector(m)
+	client := nsx.NewClusterClient(connector)
+	clusterConfig, err := client.Get()
+	if err != nil {
+		return fmt.Errorf("Failed to read Nsxt Manager Cluster: %s", err)
+	}
+	nodeInfo := clusterConfig.Nodes
+	var nodes []NsxClusterNode
+	hostIP, err := getHostIPFromClusterConfig(m, clusterConfig)
+	isIPv4 := (net.ParseIP(hostIP)).To4() != nil
+	if err != nil {
+		return err
+	}
+	for _, node := range nodeInfo {
+		ip := getIPFromNodeInfo(node, isIPv4)
+		if *ip != hostIP {
+			id := node.NodeUuid
+			clusterNode := NsxClusterNode{
+				ID:        id,
+				IPAddress: ip,
+			}
+			nodes = append(nodes, clusterNode)
+		}
+
+	}
+	d.Set("revision", clusterConfig.Revision)
+	setNodesInSchema(d, nodes)
+	return nil
+}
+
+func getIPFromNodeInfo(node nsxModel.ClusterNodeInfo, isIPv4 bool) *string {
+	// After join node into cluster, apiListen address may take some time to become available
+	// if it's not available, we retrieve node ip from nodeEntityInfo
+	var ip *string
+	apiListenAddr := node.ApiListenAddr
+	if apiListenAddr != nil {
+		if isIPv4 {
+			ip = apiListenAddr.IpAddress
+		} else {
+			ip = apiListenAddr.Ipv6Address
+		}
+	} else {
+		nodeEntities := node.Entities
+		for _, entity := range nodeEntities {
+			if *entity.Port == 443 {
+				if isIPv4 {
+					ip = entity.IpAddress
+				} else {
+					ip = entity.Ipv6Address
+				}
+				break
+			}
+		}
+	}
+	return ip
+}
+
+func resourceNsxtManagerClusterUpdate(d *schema.ResourceData, m interface{}) error {
+	connector := getPolicyConnector(m)
+	client := nsx.NewClusterClient(connector)
+	clusterConfig, err := client.Get()
+	if err != nil {
+		return fmt.Errorf("Failed to read Nsxt Manager Cluster: %s", err)
+	}
+	clusterNodeInfo := clusterConfig.Nodes
+	// get Id of nodes already in cluster from clusterConfig
+	var clusterNodeIDs []string
+	for _, nodeInfo := range clusterNodeInfo {
+		id := nodeInfo.NodeUuid
+		clusterNodeIDs = append(clusterNodeIDs, *id)
+	}
+
+	clusterID, certSha256Thumbprint, hostIP, err := getClusterInfoFromHostNode(d, m)
+	isIPv4 := (net.ParseIP(hostIP)).To4() != nil
+	nodes := getClusterNodesFromSchema(d)
+	var nodesIDFromSchema []string
+	var nodesToAdd []NsxClusterNode
+	var nodesToRemove []string
+	for _, node := range nodes {
```

Comment:
You can get node changes from the schema with `oldNodes, newNodes := d.GetChange("node")`